### PR TITLE
esp32/modsocket.c: initialize tcp when modsocket loads (fix for #26)

### PR DESCRIPTION
**Pull request by nickzoic**
This is just a simple mod to initialize tcp stack when modsocket loads, to avoid confusion.
At the moment you can't really do much with sockets unless network is configured, admittedly.

### CODE REVIEW

File: esp32/modsocket.c [hunk 1]
```diff
@@ -477,8 +478,20 @@ STATIC mp_obj_t esp_socket_getaddrinfo(const mp_obj_t host, const mp_obj_t port)
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_2(esp_socket_getaddrinfo_obj, esp_socket_getaddrinfo);
 
+STATIC mp_obj_t esp_socket_initialize() {
+    static int initialized = 0;
+    if (!initialized) {
+        ESP_LOGI("modsocket", "Initializing");
+        tcpip_adapter_init();
```
Comment (dpgeorge) [inline esp32/modsocket.c:485]: Won't tcpip_adapter_init() be called a second time when you import the network module?

File: esp32/mpconfigport.h [hunk 2]
```diff
@@ -176,7 +176,6 @@ extern const struct _mp_obj_module_t mp_module_network;
     { MP_OBJ_NEW_QSTR(MP_QSTR_random), (mp_obj_t)&mp_module_urandom }, \
     { MP_OBJ_NEW_QSTR(MP_QSTR_re), (mp_obj_t)&mp_module_ure }, \
     { MP_OBJ_NEW_QSTR(MP_QSTR_select), (mp_obj_t)&mp_module_uselect }, \
-    { MP_OBJ_NEW_QSTR(MP_QSTR_socket), (mp_obj_t)&mp_module_usocket }, \
```
Comment (dpgeorge) [inline esp32/mpconfigport.h]: To be compatible with other MicroPython's, we should keep the usocket name, and also the socket weak link.  That means you can import usocket or socket and it'll work.  And you can override usocket if you need to.

Comment (dpgeorge) [reply esp32/mpconfigport.h]: I mean you can override *socket*, by providing socket.py in your path (which could then import usocket and build upon its methods).

File: esp32/modsocket.c (hunk 1 above)
Comment (nickzoic) [reply esp32/modsocket.c:485]: Yep, but it `tcpip_adapter_init` checks if it has been called already and if so does nothing.  As an alternative, we could always init the stack on startup, but it seemed like it'd be better to wait until it was actually needed.

File: esp32/mpconfigport.h (hunk 2 above)
Comment (nickzoic) [reply esp32/mpconfigport.h]: Ah, oops, I got confused with this one and thought I'd just made a mistake ... so I made it the same as network and things started working, then I forgot all about it.

Now, looking at it, I remember what's going on: with socket in WEAK_LINKS the __init__ function doesn't get called if it is imported as 'socket', only as 'usocket'.
```
>>> import socket
>>> import usocket
I (110348) modsocket: Initializing
```
This is caused by `py/objmodule.c:mp_module_get()` which only checks `mp_module_builtin_table` and not `mp_builtin_module_weak_links_table` when deciding if it should run the module's `__init__`.  Should I put in a PR for that against the main micropython project?

**Comment by nickzoic** [comment]
(Note: the outdated changes to mpconfigport.h include some comments on running module's __init__ ... for this patch to work, for the moment you have to `import usocket` rather than `import socket`.)

**Comment by dpgeorge** [comment]
Ok, thanks, I squashed, rebased and merged this in 8a80644f75ee85a590a13fbdd3ab27b6a6996900.  And in c7830f6fc3d8925a7cab7df1cc0f73cb12d82dcd I added a small hack to call `__init__` on the module when imported via its weak link.  That latter patch needs to be reworked, eventually, but for now it works ok.

**Pull request closed without merging**